### PR TITLE
Created simple CRUD app using Spring

### PR DESCRIPTION
**Pull request by MaxSasha**

### CODE REVIEW

File: src/main/java/com/maxsasha/api/dto/UserDto.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Getter
+@Setter
```
Comment (AN0NIM17) [inline src/main/java/com/maxsasha/api/dto/UserDto.java:14]: you can replace this strings with `@Data`, same

File: src/main/java/com/maxsasha/api/dto/UserDto.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+public class UserDto {
+	@Setter(AccessLevel.NONE)
```
Comment (AN0NIM17) [inline src/main/java/com/maxsasha/api/dto/UserDto.java:16]: remove this, sometimes this setter is used

File: src/main/java/com/maxsasha/api/transformer/UserTransformer.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		for(User item : users)
+		{
+			transformList.add((UserDto.builder()
+					.id(item.getId())
+					.name(item.getName())
+					.email(item.getEmail())
+					.build()));
+		}
+		return transformList;
+	}
```
Comment (AN0NIM17) [inline src/main/java/com/maxsasha/api/transformer/UserTransformer.java:38]: Try `return users.stream().map(m -> transform(m)).collect(Collectors.toList());`
Examples: https://www.geeksforgeeks.org/stream-map-java-examples/

File: src/main/java/com/maxsasha/entity/User.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Getter
+@Setter
```
Comment (AN0NIM17) [inline src/main/java/com/maxsasha/entity/User.java:17]: You can use `@Data`

File: src/main/java/com/maxsasha/entity/User.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+@Document("users")
+public class User {
+	@Setter(AccessLevel.NONE)
```
Comment (AN0NIM17) [inline src/main/java/com/maxsasha/entity/User.java:20]: remove this

File: src/main/java/com/maxsasha/service/UserService.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.maxsasha.db.repository.UserRepository;
+import com.maxsasha.entity.User;
+
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+@Service
+public class UserService {
+	private final UserRepository dbActions;
```
Comment (AN0NIM17) [inline src/main/java/com/maxsasha/service/UserService.java:16]: I recommend rename `dbActions` to `userRepository`

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+@RequestMapping("/api")
+@RequiredArgsConstructor
+public class UserController {
+	private final UserService service;
+
+	@GetMapping("/users")
+	public List<UserDto> getUsers() {
+		log.info("Received request to get users ");
+		List<UserDto> users = UserTransformer.transform(service.getUsers());
+		log.info("Request to get users successfully completed. Users count:{}", users.size());
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:36]: remove this log. It doesn't make sence

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+@RestController
+@RequestMapping("/api")
+@RequiredArgsConstructor
+public class UserController {
+	private final UserService service;
+
+	@GetMapping("/users")
+	public List<UserDto> getUsers() {
+		log.info("Received request to get users ");
+		List<UserDto> users = UserTransformer.transform(service.getUsers());
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:35]: return result, don't save it to variabel

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		log.info("Request to get users successfully completed. Users count:{}", users.size());
+		return users;
+	}
+
+	@PostMapping("/users")
+	public void create(@RequestBody UserDto userDto) {
+		log.info("Received request to create user");
+		service.create(UserTransformer.transform(userDto));
+		log.info("Request to create successfuly completed with info: name:{} , email:{}", userDto.getName(),userDto.getEmail());
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:44]: remove it

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	public List<UserDto> getUsers() {
+		log.info("Received request to get users ");
+		List<UserDto> users = UserTransformer.transform(service.getUsers());
+		log.info("Request to get users successfully completed. Users count:{}", users.size());
+		return users;
+	}
+
+	@PostMapping("/users")
+	public void create(@RequestBody UserDto userDto) {
+		log.info("Received request to create user");
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:42]: Add logging of userDTO info on log message

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+		service.create(UserTransformer.transform(userDto));
+		log.info("Request to create successfuly completed with info: name:{} , email:{}", userDto.getName(),userDto.getEmail());
+	}
+
+	@PutMapping("/users/{id}")
+	public ResponseEntity<Object> put(@PathVariable String id, @RequestBody UserDto userDto) {
+		log.info("Received request to update user id:{}", id);
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:49]: WTF? Rewrite this method. You don't need to check if user is present or not and do all that actions

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+			log.info("Request to update successfuly completed with info: user not found");
+			return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+		}
+	}
+
+	@DeleteMapping("/users/{id}")
+	public void delete(@PathVariable("id") String id) {
+		log.info("Received request to delete user id:{}", id);
+		service.delete(id);
+		log.info("Request to delete successfuly completed");
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:67]: Remove this log

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@RestController
+@RequestMapping("/api")
+@RequiredArgsConstructor
+public class UserController {
+	private final UserService service;
+
+	@GetMapping("/users")
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:32]: move `users` uri to `RequestMapping` level, as you use it in all API's

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+	public List<UserDto> getUsers() {
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:33]: You can't work with objects, selecting all records. you should use pagination for it. Rewrite it with pagination

File: src/main/java/com/maxsasha/api/transformer/UserTransformer.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+	
+	public static UserDto transform(User user) {
+		return UserDto.builder()
+				.id(user.getId())
+				.name(user.getName())
+				.email(user.getEmail())
+				.build();
+	}
+	public static List<UserDto> transform(List<User> users)
+	{
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/transformer/UserTransformer.java:27]: Combination `Ctrl + F5` will format your code. So pls use it on future

File: src/main/java/com/maxsasha/entity/User.java [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.data.mongodb.core.mapping.Document;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Comment (koropatva) [inline src/main/java/com/maxsasha/entity/User.java:13]: Do you really needs `Builder` for `User` entity?

File: src/main/resources/application.properties [hunk 17]
```diff
@@ -0,0 +1,20 @@
+spring.data.mongodb.auto-index-creation=true
```
Comment (koropatva) [inline src/main/resources/application.properties:1]: Are you really need all those parameters on property file?

**Review by koropatva — CHANGES_REQUESTED** [review]

File: src/main/java/com/maxsasha/entity/User.java (hunk 16 above)
Comment (MaxSasha) [reply src/main/java/com/maxsasha/entity/User.java:13]: I don't need it critically, I use it once for `User` and `UserDto` in `UserTransformer`, so I can rewrite my `UserTransformer` methods using class constructors

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 18]
```diff
@@ -24,46 +27,43 @@
...
-	public List<UserDto> getUsers() {
+	@GetMapping
+	public ResponseEntity<Map<String, Object>> getUsers(@RequestParam(defaultValue = "0") int page,
+			@RequestParam(defaultValue = "3") int size) {
 		log.info("Received request to get users ");
-		List<UserDto> users = UserTransformer.transform(service.getUsers());
-		log.info("Request to get users successfully completed. Users count:{}", users.size());
-		return users;
+		Page<User> pageUsers = service.getUsers(PageRequest.of(page, size));
+		Map<String, Object> response = new HashMap<>();
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:40]: If you want to return some data structure, you should create object, populate fields and return object. Spring will automatic convert object to JSON using Jackson library

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 19]
```diff
@@ -24,46 +27,43 @@
...
 @RequiredArgsConstructor
 public class UserController {
 	private final UserService service;
 
-	@GetMapping("/users")
-	public List<UserDto> getUsers() {
+	@GetMapping
+	public ResponseEntity<Map<String, Object>> getUsers(@RequestParam(defaultValue = "0") int page,
+			@RequestParam(defaultValue = "3") int size) {
 		log.info("Received request to get users ");
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:38]: As you put page and size as parameters you should log those data

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 20]
```diff
@@ -24,46 +27,43 @@
...
+		response.put("users", UserTransformer.transform(pageUsers.getContent()));
+		response.put("Current page", pageUsers.getNumber());
+		response.put("Items count", pageUsers.getTotalElements());
+		response.put("Pages сount", pageUsers.getTotalPages());
+		return new ResponseEntity<>(response, HttpStatus.OK);
 	}
 
-	@PostMapping("/users")
+	@PostMapping
 	public void create(@RequestBody UserDto userDto) {
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:49]: When you create something you should return this object that will contains created id, in this way FE will know ID of created object

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 21]
```diff
@@ -24,46 +27,43 @@
...
-			log.info("Request to update successfuly completed with info: id:{},  name:{}, email:{}", id,
-					userDto.getName(), userDto.getEmail());
-			return ResponseEntity.status(HttpStatus.OK).body(updatedUser);
-		} else {
-			log.info("Request to update successfuly completed with info: user not found");
-			return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
-		}
+		log.info("Received request to update user with info id: {}, name: {},email: {} ", id, userDto.getName(),
+				userDto.getEmail());
+		userDto.setId(id);
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:59]: move the population of id into transformer.

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 22]
```diff
@@ -24,46 +27,43 @@
...
+		UserDto updatedUser = UserTransformer.transform(service.edit(UserTransformer.transform(userDto)));
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:60]: You don't need to make variable and populate fields for ResponseEntity, just return result and spring will populate those fields in auth way

File: src/main/java/com/maxsasha/api/transformer/UserTransformer.java [hunk 23]
```diff
@@ -9,22 +9,14 @@
 public class UserTransformer {
 
 	public static User transform(UserDto dto) {
-		return User.builder()
-				.id(dto.getId())
-				.name(dto.getName())
-				.email(dto.getEmail())
-				.build();
+		return new User(dto.getId(),dto.getName(),dto.getEmail());
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/transformer/UserTransformer.java:12]: Use builder for fields population

File: src/main/java/com/maxsasha/api/transformer/UserTransformer.java [hunk 24]
```diff
@@ -9,22 +9,14 @@
...
 	}
-	
+
 	public static UserDto transform(User user) {
-		return UserDto.builder()
-				.id(user.getId())
-				.name(user.getName())
-				.email(user.getEmail())
-				.build();
+		return new UserDto(user.getId(),user.getName(),user.getEmail());
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/transformer/UserTransformer.java:16]: same here

**Review by koropatva — CHANGES_REQUESTED** [review]

**Review by koropatva — CHANGES_REQUESTED** [review]

File: src/main/java/com/maxsasha/api/transformer/UserTransformer.java [hunk 25]
```diff
@@ -3,28 +3,22 @@
 import java.util.List;
 import java.util.stream.Collectors;
 
+import org.springframework.stereotype.Component;
+
 import com.maxsasha.api.dto.UserDto;
 import com.maxsasha.entity.User;
 
+@Component
```
Comment (MaxSasha) [inline src/main/java/com/maxsasha/api/transformer/UserTransformer.java:11]: I forgot to remove this annotation

File: src/main/java/com/maxsasha/api/controller/UserController.java [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+		log.info("Received request to get users. Page size: {}, current Page: {}", size, page);
+		return service.getUsers(PageRequest.of(page, size));
+	}
+
+	@PostMapping
+	public UserDto create(@RequestBody UserDto userDto) {
+		log.info("Received request to create user with info: name: {}, email: {}", userDto.getName(),
+				userDto.getEmail());
+		User user = service.create(transform(userDto,userDto.getId()));
+		return transform(user,user.getId());
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/controller/UserController.java:42]: You can fetch userId from User on transformer, so add this new method to transform

File: src/main/java/com/maxsasha/api/transformer/UserTransformer.java (hunk 25 above)
Comment (koropatva) [reply src/main/java/com/maxsasha/api/transformer/UserTransformer.java:11]: you have possibility to remove it

**Review by koropatva — CHANGES_REQUESTED** [review]

File: src/main/java/com/maxsasha/api/dto/StudentDto.java [hunk 27]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Data
+public class StudentDto {
+    private String id;
+    private String name;
+	private String address;
```
Comment (koropatva) [inline src/main/java/com/maxsasha/api/dto/StudentDto.java:15]: ?

**Review by koropatva — CHANGES_REQUESTED** [review]

**Pull request merged**